c++/src/capnp/rpc.c++

```diff
@@ -3134,9 +3221,26 @@ private:
         // ahead and delete it from the table.
         questions.erase(ret.getAnswerId(), *question);
       }
-
     } else {
-      KJ_FAIL_REQUIRE("Invalid question ID in Return message.") { return; }
+      auto questionId = ret.getAnswerId();
+      if (questionId >= 1 << 31) {
```

Comment:
At present this branch will never be reached because the `questions.isHigh(questionId)` check earlier on checks for exactly the same condition. This was added late last year to support the `onlyPromisePipeline` flag. When this flag is used, the caller is saying that the callee need not send any `Return` message because the call will only be used to perform promise pipelining. We end up with a similar situation where we don't know if we'll ever receive a `Return` so we need to be able to handle it without keeping any memory of the ID. So I used the same high-bit trick. Unfortunately, the semantics are slightly different than realtime messages: a `Finish` message is still sent, when the caller is done pipelining. Hence, when a caller receives a `Return` message with a question ID with the high bit set, it simply ignores the message, because it expects that a `Finish` will be sent whenever the pipeline is dropped. But for realtime messages, different behavior is needed: the caller must _always_ send a `Finish` message.

Unfortunately I think what we have to do here is split the "high ID" space further. I think we should use the _low_ bit to distinguish between these two modes. That way, we still only need one counter. The counter increments by 2 for each message sent, and then the low bit is set to 1 for realtime or 0 for onlyPromisePipelining.